File: app/src/main/java/com/ladajules/notflix/ui/auth/SignupActivity.java
package com.ladajules.notflix.ui.auth;

import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.ladajules.notflix.data.model.Profile;
import com.ladajules.notflix.data.model.User;
import com.ladajules.notflix.databinding.ActivitySignupBinding;
import com.ladajules.notflix.ui.onboarding.OnboardingActivity;
import com.ladajules.notflix.utils.Constants;
import com.ladajules.notflix.utils.Extensions;
import com.ladajules.notflix.utils.PreferenceManager;
import com.ladajules.notflix.utils.ValidationUtils;

import java.util.UUID;

public class SignupActivity extends AppCompatActivity {

    private ActivitySignupBinding binding;
    private PreferenceManager preferenceManager;
    private FirebaseAuth mAuth;
    private FirebaseFirestore mFirestore;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignupBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        mAuth = FirebaseAuth.getInstance();
        mFirestore = FirebaseFirestore.getInstance();
        preferenceManager = new PreferenceManager(this);

        setupUI();
        setupListeners();
        setupPasswordStrengthIndicator();
    }

    private void setupUI() {
        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        );
        getWindow().setStatusBarColor(Color.TRANSPARENT);
    }

    private void setupListeners() {
        binding.btnSignUp.setOnClickListener(v -> {
            Extensions.hideKeyboard(this);
            validateAndSignUp();
        });

        binding.tvSignIn.setOnClickListener(v -> finish());

        binding.ibBack.setOnClickListener(v -> finish());
    }

    private void setupPasswordStrengthIndicator() {
        binding.etPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                String password = s.toString();
                if (password.isEmpty()) {
                    binding.passwordStrengthIndicator.setVisibility(View.GONE);
                    return;
                }

                ValidationUtils.PasswordStrength strength = ValidationUtils.getPasswordStrength(password);
                binding.passwordStrengthIndicator.setVisibility(View.VISIBLE);
                binding.tvPasswordStrength.setText("Password Strength: " + strength.label);

                int color = Extensions.getColorCompat(SignupActivity.this, strength.color);
                binding.tvPasswordStrength.setTextColor(color);
                binding.progressPasswordStrength.setProgressTintList(ColorStateList.valueOf(color));

                switch (strength) {
                    case WEAK:
                        binding.progressPasswordStrength.setProgress(33);
                        break;
                    case MEDIUM:
                        binding.progressPasswordStrength.setProgress(66);
                        break;
                    case STRONG:
                        binding.progressPasswordStrength.setProgress(100);
                        break;
                }
            }
        });
    }

    private void validateAndSignUp() {
        String name = binding.etName.getText().toString().trim();
        String email = binding.etEmail.getText().toString().trim();
        String password = binding.etPassword.getText().toString().trim();
        String confirmPassword = binding.etConfirmPassword.getText().toString().trim();

        binding.tilName.setError(null);
        binding.tilEmail.setError(null);
        binding.tilPassword.setError(null);
        binding.tilConfirmPassword.setError(null);

        ValidationUtils.ValidationResult nameValidation = ValidationUtils.validateName(name);
        if (!nameValidation.isValid()) {
            binding.tilName.setError(nameValidation.getErrorMessage());
            return;
        }

        ValidationUtils.ValidationResult emailValidation = ValidationUtils.validateEmail(email);
        if (!emailValidation.isValid()) {
            binding.tilEmail.setError(emailValidation.getErrorMessage());
            return;
        }

        ValidationUtils.ValidationResult passwordValidation = ValidationUtils.validatePassword(password);
        if (!passwordValidation.isValid()) {
            binding.tilPassword.setError(passwordValidation.getErrorMessage());
            return;
        }

        ValidationUtils.ValidationResult confirmPasswordValidation = ValidationUtils.validateConfirmPassword(password, confirmPassword);
        if (!confirmPasswordValidation.isValid()) {
            binding.tilConfirmPassword.setError(confirmPasswordValidation.getErrorMessage());
            return;
        }

        signUp(name, email, password);
    }

    private void signUp(String name, String email, String password) {
        showLoading(true);

        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        String userId = mAuth.getCurrentUser().getUid();
                        createUserInFirestore(userId, name, email);
                    } else {
                        showLoading(false);
                        Extensions.showToast(this, "Registration failed: " + task.getException().getMessage());
                    }
                });
    }

    private void createUserInFirestore(String userId, String name, String email) {
        User user = new User(
                userId,
                name,
                email,
                false,
                System.currentTimeMillis(),
                System.currentTimeMillis()
        );

        mFirestore.collection(Constants.USERS_COLLECTION)
                .document(userId)
                .set(user.toMap())
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        createDefaultProfile(userId, name);
                    } else {
                        showLoading(false);
                        Extensions.showToast(this, "Failed to create user document.");
                    }
                });
    }

    private void createDefaultProfile(String userId, String name) {
        String profileId = UUID.randomUUID().toString();
        Profile defaultProfile = new Profile(
                profileId,
                userId,
                name,
                "default",
                System.currentTimeMillis()
        );

        mFirestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .set(defaultProfile.toMap())
                .addOnCompleteListener(task -> {
                    showLoading(false);
                    if (task.isSuccessful()) {
                        preferenceManager.setLoggedIn(true);
                        preferenceManager.setUserId(userId);
                        preferenceManager.setRememberMe(true);
                        preferenceManager.setSelectedProfileId(profileId);

                        Extensions.showToast(this, "Account created successfully!");
                        navigateToOnboarding();
                    } else {
                        Extensions.showToast(this, "Failed to create default profile.");
                    }
                });
    }

    private void navigateToOnboarding() {
        Intent intent = new Intent(this, OnboardingActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
        finish();
    }

    private void showLoading(boolean isLoading) {
        binding.progressBar.setVisibility(isLoading ? View.VISIBLE : View.GONE);
        binding.btnSignUp.setEnabled(!isLoading);
        binding.etName.setEnabled(!isLoading);
        binding.etEmail.setEnabled(!isLoading);
        binding.etPassword.setEnabled(!isLoading);
        binding.etConfirmPassword.setEnabled(!isLoading);
    }
}

File: app/src/main/java/com/ladajules/notflix/ui/auth/SignupActivity.kt
package com.ladajules.notflix.ui.auth

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.ladajules.notflix.R
import com.ladajules.notflix.data.model.User
import com.ladajules.notflix.data.repository.AuthRepository
import com.ladajules.notflix.data.repository.UserRepository
import com.ladajules.notflix.databinding.ActivitySignupBinding
import com.ladajules.notflix.ui.onboarding.OnboardingActivity
import com.ladajules.notflix.ui.profile.ProfileSelectionActivity
import com.ladajules.notflix.utils.PreferenceManager
import com.ladajules.notflix.utils.ValidationUtils
import com.ladajules.notflix.utils.getColorCompat
import com.ladajules.notflix.utils.hideKeyboard
import com.ladajules.notflix.utils.showToast
import kotlinx.coroutines.launch

class SignupActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySignupBinding
    private lateinit var authRepository: AuthRepository
    private lateinit var userRepository: UserRepository
    private lateinit var preferenceManager: PreferenceManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySignupBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Initialize repositories
        authRepository = AuthRepository()
        userRepository = UserRepository()
        preferenceManager = PreferenceManager(this)

        setupUI()
        setupListeners()
        setupPasswordStrengthIndicator()
    }

    private fun setupUI() {
        // Make status bar transparent
        window.decorView.systemUiVisibility = (
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                        View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                )
        window.statusBarColor = android.graphics.Color.TRANSPARENT
    }

    private fun setupListeners() {
        // Sign Up Button
        binding.btnSignUp.setOnClickListener {
            hideKeyboard()
            validateAndSignUp()
        }

        // Sign In Link
        binding.tvSignIn.setOnClickListener {
            finish() // Go back to Login
        }

        // Back Button
        binding.ibBack.setOnClickListener {
            finish()
        }
    }

    private fun setupPasswordStrengthIndicator() {
        binding.etPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable?) {
                val password = s.toString()
                if (password.isEmpty()) {
                    binding.passwordStrengthIndicator.visibility = View.GONE
                    return
                }

                val strength = ValidationUtils.getPasswordStrength(password)
                binding.passwordStrengthIndicator.visibility = View.VISIBLE
                binding.tvPasswordStrength.text = "Password Strength: ${strength.label}"

                // Update indicator color and progress
                val color = getColorCompat(strength.color)
                binding.tvPasswordStrength.setTextColor(color)
                binding.progressPasswordStrength.progressTintList =
                    android.content.res.ColorStateList.valueOf(color)

                when (strength) {
                    ValidationUtils.PasswordStrength.WEAK -> binding.progressPasswordStrength.progress = 33
                    ValidationUtils.PasswordStrength.MEDIUM -> binding.progressPasswordStrength.progress = 66
                    ValidationUtils.PasswordStrength.STRONG -> binding.progressPasswordStrength.progress = 100
                }
            }
        })
    }

    private fun validateAndSignUp() {
        val name = binding.etName.text.toString().trim()
        val email = binding.etEmail.text.toString().trim()
        val password = binding.etPassword.text.toString().trim()

        // Clear previous errors
        binding.tilName.error = null
        binding.tilEmail.error = null
        binding.tilPassword.error = null

        // Validate name
        val nameValidation = ValidationUtils.validateName(name)
        if (!nameValidation.isValid) {
            binding.tilName.error = nameValidation.errorMessage
            return
        }

        // Validate email
        val emailValidation = ValidationUtils.validateEmail(email)
        if (!emailValidation.isValid) {
            binding.tilEmail.error = emailValidation.errorMessage
            return
        }

        // Validate password
        val passwordValidation = ValidationUtils.validatePassword(password)
        if (!passwordValidation.isValid) {
            binding.tilPassword.error = passwordValidation.errorMessage
            return
        }

        // Proceed with sign up
        signUp(name, email, password)
    }

    private fun signUp(name: String, email: String, password: String) {
        showLoading(true)

        lifecycleScope.launch {
            // First, create Firebase Auth account
            when (val result = authRepository.signUp(email, password)) {
                is AuthRepository.AuthResult.Success -> {
                    // Create user document in Firestore
                    val user = User(
                        id = result.userId,
                        name = name,
                        email = email,
                        hasCompletedOnboarding = false,
                        createdAt = System.currentTimeMillis(),
                        lastLoginAt = System.currentTimeMillis()
                    )

                    val userResult = userRepository.createUser(user)

                    if (userResult.isSuccess) {
                        // Save user session
                        preferenceManager.isLoggedIn = true
                        preferenceManager.userId = result.userId
                        preferenceManager.rememberMe = true // Auto remember on signup

                        showLoading(false)
                        showToast("Account created successfully!")

                        // Navigate to Profile Selection
                        navigateToOnboarding()
                    } else {
                        showLoading(false)
                        showToast("Failed to create user profile. Please try again.")
                    }
                }
                is AuthRepository.AuthResult.Error -> {
                    showLoading(false)
                    showToast(result.message)
                }
            }
        }
    }

    private fun navigateToOnboarding() {
        val intent = Intent(this, OnboardingActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
        finish()
    }

    private fun showLoading(isLoading: Boolean) {
        binding.progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
        binding.btnSignUp.isEnabled = !isLoading
        binding.etName.isEnabled = !isLoading
        binding.etEmail.isEnabled = !isLoading
        binding.etPassword.isEnabled = !isLoading
    }
}
